pkg/agent/nodeportlocal/npl_agent_test.go

```diff
@@ -0,0 +1,409 @@
+	assert.False(t, testData.portTable.RuleExists(defaultPodIP, defaultPort))
+}
+
+// TestSvcRemoveAnnotation is the same as TestSvcUpdateAnnotation, but it deletes the NPL enabled
+// annotation, instead of setting its value to false.
+func TestSvcRemoveAnnotation(t *testing.T) {
+	testData, testSvc, testPod := setUpWithTestServiceAndPod(t)
+	defer testData.tearDown()
+
+	testSvc.Annotations = nil
+	testData.updateServiceOrFail(testSvc)
+
+	_, err := testData.pollForPodAnnotation(testPod.Name, false)
+	require.NoError(t, err, "Poll for annotation check failed")
+	assert.False(t, testData.portTable.RuleExists(defaultPodIP, defaultPort))
+}
+
+// TestSvcUpdateSelector updates the Service selector so that it no longer selects the Pod, and
+// verifies that the Pod annotation is removed.
+func TestSvcUpdateSelector(t *testing.T) {
+	testData, testSvc, testPod := setUpWithTestServiceAndPod(t)
+	defer testData.tearDown()
+
+	testSvc.Spec.Selector = map[string]string{"foo": "invalid-selector"}
+	testData.updateServiceOrFail(testSvc)
+
+	_, err := testData.pollForPodAnnotation(testPod.Name, false)
+	require.NoError(t, err, "Poll for annotation check failed")
+	assert.False(t, testData.portTable.RuleExists(defaultPodIP, defaultPort))
+}
+
+// TestPodUpdateSelectorLabel updates the Pod's labels so that the Pod is no longer selected by the
+// Service. It then verifies that the Pod's NPL annotation is removed and that the port table is
+// updated.
+func TestPodUpdateSelectorLabel(t *testing.T) {
+	testData, _, testPod := setUpWithTestServiceAndPod(t)
+	defer testData.tearDown()
+
+	testPod.Labels = map[string]string{"invalid-label-key": defaultAppSelectorVal}
+	testData.updatePodOrFail(testPod)
+
+	_, err := testData.pollForPodAnnotation(testPod.Name, false)
+	require.NoError(t, err, "Poll for annotation check failed")
+	assert.False(t, testData.portTable.RuleExists(defaultPodIP, defaultPort))
+}
+
+// TestSvcDelete deletes the Service. It then verifies that the Pod's NPL annotation is removed and
+// that the port table is updated.
+func TestSvcDelete(t *testing.T) {
+	testData, testSvc, testPod := setUpWithTestServiceAndPod(t)
+	defer testData.tearDown()
+
+	err := testData.k8sClient.CoreV1().Services(defaultNS).Delete(context.TODO(), testSvc.Name, metav1.DeleteOptions{})
+	require.NoError(t, err, "Service deletion failed")
+	t.Logf("successfully deleted Service: %s", testSvc.Name)
+
+	_, err = testData.pollForPodAnnotation(testPod.Name, false)
+	require.NoError(t, err, "Poll for annotation check failed")
+	assert.False(t, testData.portTable.RuleExists(defaultPodIP, defaultPort))
+}
+
+// TestPodUpdate verifies that any update in the Pod container port is reflected in Pod annotation
+// and local port table.
+func TestPodUpdate(t *testing.T) {
+	testData, _, testPod := setUpWithTestServiceAndPod(t)
+	defer testData.tearDown()
+
+	newPort := 8080
+	testPod.Spec.Containers[0].Ports[0].ContainerPort = int32(newPort)
+	testData.updatePodOrFail(testPod)
```

Comment:
Not introduced by this PR, but updating container port is not a supported operation so can never happen:
```
spec: Forbidden: pod updates may not change fields other than `spec.containers[*].image`, `spec.initContainers[*].image`, `spec.activeDeadlineSeconds` or `spec.tolerations` (only additions to existing tolerations)
```
Is there logic in main code to handle this update? @monotosh-avi @shubhamavi

Reply:
Thanks Quan. Looks like the fake K8s client does not have these restrictions, hence why the test succeeds. The NPL implementation handle Pod updates regardless of which fields have been modified. Looks like it can be improved to ignore all spec updates (we only care about label updates). We can make that change in a future PR as a small improvement.

I will update the PR to remove this test, since it is invalid.